guides/v2.3/rest/bulk-endpoints.md

```diff
@@ -9,6 +9,12 @@ functional_areas:
 
 Bulk API endpoints differ from other REST endpoints in that they combine multiple calls of the same type into an array and execute them as a single request. The endpoint handler splits the array into individual entities and writes them as separate messages to the message queue.
 
+{: .bs-callout .bs-callout-warning}
+By default, Magento uses RabbitMQ (AMQP) for processing message queues. Before using Bulk operations, you should see [RabbitMQ]({{ page.baseurl }}/install-gde/prereq/install-rabbitmq.html).
+
+{:.bs-callout .bs-callout-info}
+Alternatively you can use CRON for processing message queues without RabbitMQ installed, see [Manage message queues]({{ page.baseurl }}/config-guide/mq/manage-message-queues.html).
```

Comment:
Correct me if I'm wrong, but I'm almost 100% convinced that cron should always be used, even if you use RabbitMQ, otherwise there won't be any consumers running and messages will only end up in the queue but will not get processed.
There seems to be a lot of misinformation out there around this fact.

The exception is if you use something like [supervisor to manage consumers](https://devdocs.magento.com/guides/v2.3/config-guide/mq/manage-message-queues.html#process-management), but that has nothing to do with RabbitMQ vs MysqlMQ...

Reply:
Hi @hostep base on @rogyar [previous](https://github.com/magento/devdocs/pull/5569#discussion_r331595732) explanation and first paragraph of [
Manage message queues
](https://devdocs.magento.com/guides/v2.3/config-guide/mq/manage-message-queues.html#process-management) - "If you don’t want to implement the RabbitMQ solution, you can manage message queues with cron jobs (or an external process manager) and the command line to ensure that consumers are retrieving messages." - Is not necessary have the job 'consumers_runner' enabled if you use RabbitMQ to manage queues.

Reply:
Well, that's what I mean by
> There seems to be a lot of misinformation out there around this fact.

😉 

Anyway, just to make sure I know what I'm talking about, I just tested this again to be very sure, here are some graphs from the RabbitMQ management interface:
<img width="875" alt="Screenshot 2019-10-07 at 22 11 30" src="https://user-images.githubusercontent.com/85479/66345112-a4ea0880-e94f-11e9-9b41-334c78db6ff4.png">

Between 22:06 and 22:10, I send in 3 REST calls to the bulk endpoints. No cronjobs or message queue consumers were running at that time. Notice that messages are added to the queue but stay in there and aren't getting processed.

At around 22:11, I executed a `bin/magento cron:run` which spawned some message queue consumer processes in the background. Notice that suddenly all messages are getting processed and the queue is empty again.

So using RabbitMQ by itself is not enough, you need consumers running within Magento so they can process the pending messages in the queue. Those consumers either get spawned using Magento's cronjob, you can start them manually using `bin/magento queue:consumers:start ...`, or you can use something like supervisor to manage them.

Is this enough information which you guys can use to fix all incorrect information on devdocs around this? Thanks! 🙂

Reply:
Thanks @hostep. I'll investigate it. Probably the misinformation is caused by de description of the [Manage message queues](https://devdocs.magento.com/guides/v2.3/config-guide/mq/manage-message-queues.html#process-management) that explains:

> If you don’t want to implement the RabbitMQ solution, you can manage message queues with cron jobs ...

But at [RabbitMQ](https://devdocs.magento.com/guides/v2.3/install-gde/prereq/install-rabbitmq.html) explains that Crons are necessary to run consumers:

> After you have connected Magento and RabbitMQ, you must start the message queue consumers. See Configure message queues for details.

Probably the mistake is in the [Manage message queues](https://devdocs.magento.com/guides/v2.3/config-guide/mq/manage-message-queues.html#process-management) description that cause this confusion :(

Reply:
Hi there!

Consumers can use RabbitMQ or the DB for messages. It depends on the implementation of the consumer.
But to run consumers you need to run the command `bin/magento queue:consumers:start ...`
This command can be run by the cron job or some worker (for example, you can use workers on Magento Cloud Pro instances)

Reply:
No.  I still need to review the PR, and there is also a confusion between running the consumer to process the messages, which is covered by the start consumers  and configuring the message queues.

@brunoroeder Check this topic:  https://devdocs.magento.com/guides/v2.3/extension-dev-guide/message-queues/implement-bulk.html

I also think that this link would be a better one for linking to how to start consumers:  https://devdocs.magento.com/guides/v2.3/config-guide/cli/config-cli-subcommands-queue.html

Reply:
@brunoroeder   I provided updated comments in the PR.   Short story:  Edits to note referring to start consumer content, and remove note about RabbitMQ.   Let me know if you have any questions. Thx.

Reply:
Hi @hostep , @meker12 
> Correct me if I'm wrong, but I'm almost 100% convinced that cron should always be used

Its not now. CRON really still CAN be used. But you can also disable your consumer to run by using cron. And them align your server to run consumer as a service, without basically letting Magento know that its running.

So Cronjob is really optional on this step, same us direct running on consumer process. This decision is up to developers

Reply:
Hi @nuzil, thanks for the feedback!

I agree with your comment, sorry that was a bad wording from me.
But I don't agree with the current proposal as it seems to suggest that you need to run RabbitMQ **or** need to run cronjobs for the messages to get processed.

This is not the case. You need a place to store the messages (RabbitMQ or MysqlMQ) **and** consumer processes to do something with those messages. You need both, one of them by itself is not going to work.

For bulk api, you always need RabbitMQ (at the moment of writing) and using cronjobs as an alternative makes no sense. The cronjob which is being referred to here, is the `consumer_runner` one which spawns these consumer processes.

You are correct that cronjobs aren't necessary to spawn consumer processes, you can manually run them or have some other system manage them. But you do need consumer processes running to have the messages being processed.
So they are **optional**, but are not an **alternative** as what is being suggested.

I already tried to explain it a bit better previously a bit higher up, let me quote myself:

> So using RabbitMQ by itself is not enough, you need consumers running within Magento so they can process the pending messages in the queue. Those consumers either get spawned using Magento's cronjob, you can start them manually using bin/magento queue:consumers:start ..., or you can use something like supervisor to manage them.


I'm not seeing ghosts right? I did test this many times now already ... 🙂

Reply:
Right, I guess I skip this part of your message :-)
No you are totally correct described situation 👍 , with this amount of discussions can be wrongly to define what is the last point

Reply:
> Hi @nuzil, thanks for the feedback!
> 
> I agree with your comment, sorry that was a bad wording from me.
> But I don't agree with the current proposal as it seems to suggest that you need to run RabbitMQ **or** need to run cronjobs for the messages to get processed.
> 
> This is not the case. You need a place to store the messages (RabbitMQ or MysqlMQ) **and** consumer processes to do something with those messages. You need both, one of them by itself is not going to work.
> 
> For bulk api, you always need RabbitMQ (at the moment of writing) and using cronjobs as an alternative makes no sense. The cronjob which is being referred to here, is the `consumer_runner` one which spawns these consumer processes.
> 
> You are correct that cronjobs aren't necessary to spawn consumer processes, you can manually run them or have some other system manage them. But you do need consumer processes running to have the messages being processed.
> So they are **optional**, but are not an **alternative** as what is being suggested.
> 
> I already tried to explain it a bit better previously a bit higher up, let me quote myself:
> 
> > So using RabbitMQ by itself is not enough, you need consumers running within Magento so they can process the pending messages in the queue. Those consumers either get spawned using Magento's cronjob, you can start them manually using bin/magento queue:consumers:start ..., or you can use something like supervisor to manage them.
> 
> I'm not seeing ghosts right? I did test this many times now already ... 🙂